File: feed/luci-app-invizbox/src/luci/model/cbi/dns.lua
-- Copyright 2020 InvizBox Ltd
-- https://www.invizbox.com/lic/license.txt

local cbi = require "luci.cbi"
local translate = require "luci.i18n"
local sys = require("luci.sys")

local map, dns_providers, dns_provider, dummyvalue, product_name

map = cbi.Map("wizard", translate.translate("DNS"))
map.anonymous = true
map:chain("dhcp")
map:chain("dns")

dns_providers = map:section(cbi.NamedSection, "main", "wizard", "", translate.translate("Choose your DNS provider:"))
dns_providers.addremove = false

dns_provider = dns_providers:option(cbi.ListValue, "dns_id", translate.translate("DNS Providers:"))
dns_provider:value("dhcp", "from Router (WAN)")
dns_provider.default = "dhcp"
map.uci:foreach("dns", "servers", function(section)
    dns_provider:value(section[".name"], section["name"])
end)

product_name = map.uci:get("wizard", "main", "product_name") or "InvizBox Go"

dummyvalue = dns_providers:option(cbi.DummyValue, "_provider")
dummyvalue.template = "cbi/dvalue"
dummyvalue.rawhtml  = true
dummyvalue.value = '<p class="note">'..translate.translate("This DNS is used by the "..product_name..
        " as well all devices in WiFi Extender mode.")..'</p>'

function map.on_commit()
    local dns_id = dns_provider:formvalue("main")
    if dns_id == "dhcp" then
        map.uci:delete("dhcp", "auto", "noresolv")
        map.uci:set("dhcp", "auto", "resolvfile", "/tmp/resolv.conf.auto")
        map.uci:delete("dhcp", "auto", "server")
        map.uci:delete("dhcp", "invizbox", "noresolv")
        map.uci:set("dhcp", "invizbox", "resolvfile", "/tmp/resolv.conf.auto")
        map.uci:set("dhcp", "invizbox", "server", {"/onion/172.31.1.1#9053"})
    else
        local new_dns_servers = {}
        for _, server in pairs(map.uci:get("dns", dns_id, "dns_server")) do
            table.insert(new_dns_servers, server)
        end
        map.uci:set("dhcp", "auto", "noresolv", "1")
        map.uci:delete("dhcp", "auto", "resolvfile")
        map.uci:set("dhcp", "auto", "server", new_dns_servers)
        map.uci:set("dhcp", "invizbox", "noresolv", "1")
        map.uci:delete("dhcp", "invizbox", "resolvfile")
        table.insert(new_dns_servers, "/onion/172.31.1.1#9053")
        map.uci:set("dhcp", "invizbox", "server", new_dns_servers)
    end
    map.uci:save("dhcp")
    map.uci:commit("dhcp")
end

function map.on_after_commit()
    sys.call("/etc/init.d/dnsmasq restart")
end

return map

File: feed/luci-app-invizbox/htdocs/luci-static/resources/view/invizbox/dns.js
'use strict';
'require view';
'require fs';
'require ui';
'require rpc';
'require uci';
'require form';

const dns = view.extend({
  load: function(){ return Promise.all([uci.changes(), uci.load('dns')]); },
  render: function(){
    let m, s, o;

    m = new form.Map('dns');
    m.chain('dhcp');

    s = m.section(form.NamedSection, 'main', 'dns', 'DNS');

    o = s.option(form.ListValue, 'dns_id', 'DNS Provider',
      'This DNS is used by the InvizBox as well as all devices in WiFi Extender mode.');
    o.value('dhcp', 'from Router (WAN)');
    uci.sections('dns', 'servers', function(section){ return o.value(section['.name'], section.name); });

    return m.render();
  },
  handleSaveApply: function(ev, mode) {
    return this.handleSave(ev).then(function() {
      const dns_id = document.getElementById('widget.cbid.dns.main.dns_id').value;
      if (dns_id === 'dhcp') {
        uci.set('dhcp', 'auto', 'noresolv', null);
        uci.set('dhcp', 'auto', 'resolvfile', '/tmp/resolv.conf.d/resolv.conf.auto');
        uci.set('dhcp', 'auto', 'server', null);
        uci.set('dhcp', 'invizbox', 'noresolv', null);
        uci.set('dhcp', 'invizbox', 'resolvfile', '/tmp/resolv.conf.d/resolv.conf.auto');
        uci.set('dhcp', 'invizbox', 'server', ['/onion/172.31.1.1#9053']);
      } else {
        uci.set('dhcp', 'auto', 'noresolv', '1');
        uci.set('dhcp', 'auto', 'resolvfile', null);
        uci.set('dhcp', 'auto', 'server', uci.get('dns', dns_id, 'dns_server'));
        uci.set('dhcp', 'invizbox', 'noresolv', '1');
        uci.set('dhcp', 'invizbox', 'resolvfile', null);
        uci.set('dhcp', 'invizbox', 'server',
          uci.get('dns', dns_id, 'dns_server').concat('/onion/172.31.1.1#9053'));
      }
      uci.save();
      classes.ui.changes.apply(mode == '0');
      fs.exec_direct('/etc/init.d/dnsmasq', [ 'restart' ]);
    });
  },
  handleReset: null
});
return dns;
